spec/models/services/data_consolidator_spec.rb

```diff
@@ -0,0 +1,114 @@
+end
+
+describe Maestrano::Connector::Rails::Services::DataConsolidator do
+  let!(:organization) {create(:organization, uid: 'cld-123')}
+  let(:external_entities) {
+    [
+      {
+        'amount' => 8.93,
+        'type' => 'RECEIVE',
+        'public_note' => '',
+        'transaction_date' => '2018-07-31 12:00:00 +0000',
+        'name' => 'INTEREST PAYMENT',
+        'id' => 'TX8934077941420180731',
+        'account_id' => '4518736'
+      },
+      {
+        'amount' => -26.47,
+        'type' => 'SPEND',
+        'public_note' => 'DISBURSEMENT TO MORTGAGOR',
+        'transaction_date' => '2018-01-16 12:00:00 +0000',
+        'name' => 'DISBURSEMENT TO MORTGAGOR',
+        'id' => '201801160',
+        'account_id' => '4518738'
+      },
+      {
+        'amount' => 25.0,
+        'type' => 'RECEIVE',
+        'public_note' => 'LATE CHARGE PAID',
+        'transaction_date' => '2018-01-12 12:00:00 +0000',
+        'name' => 'LATE CHARGE PAID',
+        'id' => '201801121',
+        'account_id' => '4518738'
+      }
+    ]
+  }
+
+  class BankTransactionMapper
+    extend HashMapper
+    map from('name'), to('name')
+    map from('transaction_date'), to('transaction_date')
+    map from('type'), to('type')
+    map from('public_note'), to('public_note')
+    map from('account_id'), to('account_id')
+    map from('id'), to('id')
+    map from('lines'), to('lines')
+    map from('status'), to('status')
+    map from('amount'), to('amount')
+  end
+
+  class BankTransaction < Maestrano::Connector::Rails::Entity
+    def self.id_from_external_entity_hash(entity)
+      entity['id']
+    end
+
+    def self.connec_entity_name
+      'Transaction'
+    end
+
+    # Entity name in external system
+    def self.external_entity_name
+      'Transaction'
+    end
+
+    def self.object_name_from_external_entity_hash(entity)
+      self.id_from_external_entity_hash(entity)
+    end
+
+    def self.mapper_class
+      BankTransactionMapper
+    end
+  end
+
+  describe '#consolidate_external_entities' do
+    let(:transaction) {BankTransaction.new(organization, nil, nil)}
+    let(:data_consolidator) {described_class.new(organization, transaction, {})}
+
+    context 'entity is not immutable, no entity has been pushed to connec! before' do
+      it 'returns all external entities' do
+        allow(BankTransaction).to receive(:immutable?).and_return(false)
+        entities = data_consolidator.consolidate_external_entities(external_entities, 'BankTransaction')
+        ent = process_entities_to_cmp(entities)
+        expect(ent).to eq(external_entities)
+      end
+    end
+
+    context 'entity is immutable and some entities have been pushed to connec! before' do
+      it 'returns some external entities' do
+        Maestrano::Connector::Rails::IdMap.create(external_id: 'TX8934077941420180731', external_entity: 'transaction', organization_id: organization.id, connec_entity: 'banktransaction', connec_id: 1)
```

Comment:
Same for the creation and the stub. The context setup should be in a before block.
The `it` blocks should be quite small

Especially since the immutable is repeated for the next step.

So something like this:
```ruby
context 'when entity is immutable' do
  before { allow(BankTransaction).to receive(:immutable?).and_return(true) }

  context 'when some entities have already been pushed to connec!' do
   before do
    Maestrano::Connector::Rails::IdMap.create(external_id: 'TX8934077941420180731', external_entity: 'transaction', organization_id: organization.id, connec_entity: 'banktransaction', connec_id: 1)
   end
   it 'returns some external entities' do# Or it 'only returns the new entities'
     ...
     # Only 1 expectation here which is clearer than the 3 separates
     expect(ent).to match_array(external_entities[1], external_entities[2])
   end
  end

  context 'when all entities have already been pushed to connec!' do
    before do
      ...
    end
    it 'does not return any entities'
  end
end
```